### ISSUE TITLE
False positive in CA1859

### ISSUE THREAD

**Issue by buyaa-n** [dotnet/roslyn-analyzers#6751]
### Analyzer

**Diagnostic ID**: [CA1859](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/CA1859): `Use concrete types when possible for improved performance`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://learn.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

### False positive:
Warns: `Change return type of method 'ToRecipientInfosForThisIndex' from 'System.Collections.
Generic.IEnumerable<System.Security.Cryptography.Pkcs.RecipientInfo>' to 'System.Security.Cryptography.Pkcs.KeyAgreeRecipientInfo[]' for improved performance` when one of the cases return `KeyTransRecipientInfo[]` instead of `KeyAgreeRecipientInfo[]`

```cs
        private static IEnumerable<RecipientInfo> ToRecipientInfosForThisIndex(SafeHandle pCmsgCmsRecipientInfoMemory, int index)
        {
            bool mustRelease = false;
            pCmsgCmsRecipientInfoMemory.DangerousAddRef(ref mustRelease);
            try
            {
                unsafe
                {
                    CMSG_CMS_RECIPIENT_INFO* pCMsgCmsRecipientInfo = (CMSG_CMS_RECIPIENT_INFO*)(pCmsgCmsRecipientInfoMemory.DangerousGetHandle());
                    switch (pCMsgCmsRecipientInfo->dwRecipientChoice)
                    {
                        case CMsgCmsRecipientChoice.CMSG_KEY_TRANS_RECIPIENT:
                            return new KeyTransRecipientInfo[] { new KeyTransRecipientInfo(new KeyTransRecipientInfoPalWindows(pCmsgCmsRecipientInfoMemory, index)) };

                        case CMsgCmsRecipientChoice.CMSG_KEY_AGREE_RECIPIENT:
                            {
                                CMSG_KEY_AGREE_RECIPIENT_INFO* pCmsKeyAgreeRecipientInfo = pCMsgCmsRecipientInfo->KeyAgree;
                                int numKeys = pCmsKeyAgreeRecipientInfo->cRecipientEncryptedKeys;
                                KeyAgreeRecipientInfo[] recipients = new KeyAgreeRecipientInfo[numKeys];
                                for (int subIndex = 0; subIndex < numKeys; subIndex++)
                                {
                                    recipients[subIndex] = new KeyAgreeRecipientInfo(new KeyAgreeRecipientInfoPalWindows(pCmsgCmsRecipientInfoMemory, index, subIndex));
                                }
                                return recipients;
                            }

                        default:
                            throw ErrorCode.E_NOTIMPL.ToCryptographicException();
                    }
                }
            }
            finally
            {
                if (mustRelease)
                {
                    pCmsgCmsRecipientInfoMemory.DangerousRelease();
                }
            }
        }
```
### Expected behavior
No warning
### Actual behavior
Warns `Change return type of method 'ToRecipientInfosForThisIndex' from 'System.Collections.
Generic.IEnumerable<System.Security.Cryptography.Pkcs.RecipientInfo>' to 'System.Security.Cryptography.Pkcs.KeyAgreeRecipientInfo[]' for improved performance`

CC @geeknoid

**Comment by buyaa-n** [comment]
Another possible false positive:

Warning to change the following local variables type into array `ParameterInfo[]`
``` cs
        public void M()
        {
            IList<ParameterInfo> delegateParameters = invokeMethod.GetParametersNoCopy();
            IList<ParameterInfo> targetParameters = this.GetParametersNoCopy();
            IEnumerator<ParameterInfo> delegateParameterEnumerator = delegateParameters.GetEnumerator();
            IEnumerator<ParameterInfo> targetParameterEnumerator = targetParameters.GetEnumerator();
            // further other usages of the list
        }
```
If make it an array the next row `IEnumerator<ParameterInfo> delegateParameterEnumerator = delegateParameters.GetEnumerator();` would complain, I think we should not warn for this case

**Comment by geeknoid** [comment]
@buyaa-n Could you explain the second case you show? I don't understand what's wrong with making delegateParameters into an array. It would avoid the interface dispatch used for GetEnumerator.

**Comment by buyaa-n** [comment]
The problem is after changing the row `IList<ParameterInfo> delegateParameters = invokeMethod.GetParametersNoCopy();` into array `ParameterInfo[] delegateParameters = invokeMethod.GetParametersNoCopy();`  there is a compiler error for next statement `IEnumerator<ParameterInfo> delegateParameterEnumerator = delegateParameters.GetEnumerator();`

I am not sure this is a false positive exactly, though we might want to avoid causing compiler error.

**Comment by mavasani** [comment]
@buyaa-n Feel free to open a separate issue if the second reported false positive still holds after #6777
